### DISCUSSION TITLE
CI Hooks

### DISCUSSION THREAD

**Discussion by tozonaran** [semaphoreui/semaphore#841] [Q&A]
Can anybody please help me to trigger Tasks via Bitbucket hooks?

I am trying to write a bash script to use the API however I am finding this difficult.

Thanks

Taner

**Comment by gcavalcante8808** [comment]
I think a question related to this request is the support for git/gitlab hooks as well. Should we  rename this enhancement to "CI Hooks" ?

**Comment by awartoft** [comment]
Any chance for an update on this ?

**Comment by matejkramny** [comment]
Can you propose how it would work? Is it the same as github's webhooks?

**Comment by strangeman** [comment]
I think we just need to create a plain, clear documentation about API: how to create user token, how to trigger task via API, etc.
At the current moment, we have only technical API docs and they not so user-friendly.

**Comment by thedumbtechguy** [comment]
@strangeman

I was planning on going with docker, but my few experiments show that that's more headache than it's worth. 
I'd rather go with a full ansible build process. 

I've seen that I can trigger a task via the api. 
Can you give me some pointers on how to handle the api tokens, I'd be more than happy to submit documentation on how to get the process running once I get it working on my end.

**Comment by strangeman** [comment]
@frostymarvelous here is example for curl:

```
# login to the semaphore (password should be escaped, "slashy\\pass" instead of "slashy\pass" e.g.)
curl --cookie-jar /tmp/semaphore-cookie -X POST --header 'Content-Type: application/json' --header 'Accept: application/json'  -d '{"auth": "semaphore_user","password": "semaphore_password"}' https://ansible.yourdomain.com/api/auth/login -v
#204 Responce code should be returned

# get user tokens
curl -b /tmp/semaphore-cookie https://ansible.yourdomain.com/api/user/tokens -v
# If user have no tokens, empty array will be returned

#generate a new token if user had no one
curl -b /tmp/semaphore-cookie -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' https://ansible.yourdomain.com/api/user/tokens -v
#201 Responce code should be returned

# get user tokens again
curl -b /tmp/semaphore-cookie https://ansible.yourdomain.com/api/user/tokens -v
#Should be returned something like:
#[{"id":"long_random_token_string","created":"2017-03-11T13:13:13Z","expired":false,"user_id":1}]

# use this token for launching a task or anything else
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' --header 'Authorization: Bearer long_random_token_string' -d '{"template_id": 1, "debug": false, "dry_run": false, "playbook": "", "environment": ""}' https://ansible.yourdomain.com/api/project/1/tasks -v

#alternatively you still may use cookie
curl -X POST -b /tmp/semaphore-cookie --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{"template_id": 1, "debug": false, "dry_run": false, "playbook": "", "environment": ""}' https://ansible.yourdomain.com/api/project/1/tasks -v
```

You can get Template ID  from Update Template dialog (if you use latest semaphore version).
More documentation about Semaphore API here - https://ansible-semaphore.github.io/semaphore/

**Comment by strangeman** [comment]
We call Semaphore API from Jenkins via following snippet: https://gist.github.com/strangeman/a7abc310940247730aaa24a22cdf19ba

Need to install https://wiki.jenkins-ci.org/display/JENKINS/HTTP+Request+Plugin, API Token generated via curl and stored in Jenkins Credentials.

**Comment by thedumbtechguy** [comment]
Thank you.

I'll give it a go with Gitlab and let you know what I come up with?

**Comment by strangeman** [comment]
@frostymarvelous I have no experience with GitLab, but I think you can write simple BASH/Python/Ruby script that makes POST request to the API and add it to the post-receive Git hook - https://docs.gitlab.com/ce/administration/custom_hooks.html

**Comment by thedumbtechguy** [comment]
Thanks for that tip.

Unfortunately, that won't work for us.

I am planning on building a small proxy to handle the webhook (we need
builds on accepted merges) as we currently use gitlab.com and semaphore is
behind a private network.

I'll tackle this a bit later though.
Currently busy trying to automate our infrastructure first.
